cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -0,0 +1,1722 @@
+    if ((src.find("127.0.0.1") != std::string::npos)) {
+      RETURN_NOT_OK(FromLocalHostString(src));
+    }
+    if (internal::IsLikelyUri(src)) {
+      RETURN_NOT_OK(ExtractBlobPath(src));
+    }
+    src = internal::RemoveLeadingSlash(src);
+    auto first_sep = src.find_first_of(kSep);
+    if (first_sep == 0) {
+      return Status::Invalid("Path cannot start with a separator ('", s, "')");
+    }
+    if (first_sep == std::string::npos) {
+      return AzurePath{std::string(src), std::string(src), "", {}};
+    }
+    AzurePath path;
+    path.full_path = std::string(src);
+    path.container = std::string(src.substr(0, first_sep));
+    path.path_to_file = std::string(src.substr(first_sep + 1));
+    path.path_to_file_parts = internal::SplitAbstractPath(path.path_to_file);
+    RETURN_NOT_OK(Validate(&path));
+    return path;
+  }
+
+  static Status FromLocalHostString(util::string_view& src) {
+    auto port = src.find("127.0.0.1");
+    src = src.substr(port);
+    auto first_sep = src.find_first_of(kSep);
+    src = src.substr(first_sep + 1);
+    auto sec_sep = src.find_first_of(kSep);
+    src = src.substr(sec_sep + 1);
+    return Status::OK();
+  }
+
+  // Removes scheme, host and port from the uri
+  static Status ExtractBlobPath(util::string_view& s) {
+    Uri uri;
+    RETURN_NOT_OK(uri.Parse(s.to_string()));
+    s = uri.path();
+    return Status::OK();
+  }
+
+  static Status Validate(const AzurePath* path) {
+    auto result = internal::ValidateAbstractPathParts(path->path_to_file_parts);
+    if (!result.ok()) {
+      return Status::Invalid(result.message(), " in path ", path->full_path);
+    } else {
+      return result;
+    }
+  }
+
+  AzurePath parent() const {
+    DCHECK(!path_to_file_parts.empty());
+    auto parent = AzurePath{"", container, "", path_to_file_parts};
+    parent.path_to_file_parts.pop_back();
+    parent.path_to_file = internal::JoinAbstractPath(parent.path_to_file_parts);
+    if (parent.path_to_file.empty()) {
+      parent.full_path = parent.container;
+    } else {
+      parent.full_path = parent.container + kSep + parent.path_to_file;
+    }
+    return parent;
+  }
+
+  bool has_parent() const { return !path_to_file.empty(); }
+
+  bool empty() const { return container.empty() && path_to_file.empty(); }
+
+  bool operator==(const AzurePath& other) const {
+    return container == other.container && path_to_file == other.path_to_file;
+  }
+};
+
+template <typename ObjectResult>
+std::shared_ptr<const KeyValueMetadata> GetObjectMetadata(const ObjectResult& result) {
+  auto md = std::make_shared<KeyValueMetadata>();
+  auto push = [&](std::string k, const std::string v) {
+    if (!v.empty()) {
+      md->Append(std::move(k), v);
+    }
+  };
+  for (auto prop : result) {
+    push(prop.first, prop.second);
+  }
+  return md;
+}
+
+template <typename T>
+Status InitServiceClient(std::shared_ptr<T>& client, const AzureOptions options,
+                         const std::string url) {
```

Comment:
Same here: avoid using non-const references. Here, you should probably return a `Result<std::shared_ptr<T>>` instead...